Add references and flags to constant pointers to data maps

wargio opened the pull request:
# DO NOT SQUASH

**Your checklist for this pull request**
- [ ] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [ ] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

<!-- Explain the **details** for making this change. Is a new feature implemented? What existing problem does the pull request solve? How does the pull request solve these issues? Please provide enough information so that others can review your pull request. -->

...

**Test plan**

<!-- What steps should the reviewer take to test your pull request? Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots/videos. This is your time to re-check that everything works and that you covered all the edge cases -->

...

**Closing issues**

<!-- put "closes #XXXX" in your comment to auto-close the issue that your PR fixes (if any). -->

...

wargio committed:
Add references and flags to constant pointers to data maps

ret2libc commented on librz/core/canalysis.c:
```diff
@@ -5755,6 +5755,98 @@ RZ_IPI bool rz_core_analysis_var_rename(RzCore *core, const char *name, const ch
+}
+
+/**
+ * \brief      Tries to resolve all the constant pointers and adds flags named ptr.XXXXXX
```
```suggestion
 * \brief      Tries to resolve all the constant pointers and adds flags named data.XXXXXX
```

ret2libc commented on librz/core/canalysis.c:
```diff
@@ -5755,6 +5755,98 @@ RZ_IPI bool rz_core_analysis_var_rename(RzCore *core, const char *name, const ch
+ * \param      core  The RzCore to analyze
+ */
+RZ_IPI void rz_core_analysis_resolve_pointers_to_data(RzCore *core) {
+	if (rz_str_startswith(rz_config_get(core->config, "asm.arch"), "dalvik")) {
+		// never run this for dalvik
+		return;
+	}
+
```
why?

wargio replied:
there is no need to do this with dalvik stuff since everything is resolved by default. there are no pointers in memory.

thestr4ng3r replied:
Add this as a comment to the code please

ret2libc replied:
The same might be true for other plugins as well? I don't like having special case in generic code like this, it feels hacky. Would anything happen if we just do it anyway? We should consider that someone not in rizinorg might write a plugin that we are not aware about and the non-plugin-specific code should work well nonetheless.

wargio replied:
i think many of these things can be solved once we have RzArch.

ret2libc replied:
How? I don't think It will be the magic solution for all of these special if we have around the code base

wargio replied:
just some options to say if the arch does use pointers etc.. like wasm does not either. it is a stack only based arch and the stack does not have a size. just a limit to how many values can be stored.

ret2libc replied:
Can we do that already in reanalysis plugins please? It's easier to port to rzarch if it's already in the interface rather than having to chase all ifs around

wargio replied:
what analysis plugin? i think this is the right place to add such check and filter out useless loops. tbh, this probably should be done when we analyze any instruction.

ret2libc commented on librz/core/canalysis.c:
```diff
@@ -5755,6 +5755,98 @@ RZ_IPI bool rz_core_analysis_var_rename(RzCore *core, const char *name, const ch
+			}
+
+			for (ut32 i = 0; i < block->size;) {
+				ut64 pc = block->addr + i;
+				RzAnalysisOp aop;
+				rz_analysis_op_init(&aop);
+				if (rz_analysis_op(core->analysis, &aop, pc, bytes + i, (block->size - i), RZ_ANALYSIS_OP_MASK_DISASM) < 1) {
+					rz_analysis_op_fini(&aop);
+					i += min_op_size;
+					continue;
+				}
+				i += aop.size;
```
we probably need an API to iterate over all instructions in a basicblock and that calls a callback passing each RzAnalysisOp, so that we can avoid  this thing every time.

ret2libc replied:
Btw, i was not suggesting you do this yourself in this PR:) i was just saying, feel free to ignore this comment

wargio replied:
i wanted to add this in the PR and then i was like, probably a bad idea :D KISS!!

ret2libc commented on librz/core/canalysis.c:
```diff
@@ -5755,6 +5755,98 @@ RZ_IPI bool rz_core_analysis_var_rename(RzCore *core, const char *name, const ch
+	if (address < 256 || address == UT64_MAX) {
+		return false;
+	}
+	RzBinObject *o = rz_bin_cur_object(core->bin);
+	const RzBinSection *section = o ? rz_bin_get_section_at(o, address, core->io->va) : NULL;
+	return section && section->vsize > 0 && strstr(section->name, "data");
```
what if we rely on segment instead and check for r-- or rw- permissions?

wargio replied:
because is not reliable. there are bins with data that has the EXEC flag.

ret2libc replied:
Sections are even less reliable considering that in binaries like ELF they can even be removed completely without affecting the program execution at all. 

I think we can rely on both sections and segments. If the address falls in a section with data name, consider it data. Otherwise look at segment and their permissions.

ret2libc replied:
We can consider both Rx and rw and r as data. Rwx means self modifying code and who knows what happens in that case.

wargio replied:
hmm, ok agreed.

ret2libc reviewed (commented):
nice stuff

wargio commented on librz/core/canalysis.c:
```diff
@@ -5755,6 +5755,110 @@ RZ_IPI bool rz_core_analysis_var_rename(RzCore *core, const char *name, const ch
 	return true;
 }
 
+static bool is_in_data_section(RzCore *core, const ut64 address) {
+	if (address < 256 || address == UT64_MAX) {
+		return false;
+	}
+	RzBinObject *o = rz_bin_cur_object(core->bin);
+	const RzBinSection *section = o ? rz_bin_get_section_at(o, address, core->io->va) : NULL;
+	if (!section || section->size < 2) {
+		return false;
+	}
+
+	if (strstr(section->name, "data")) {
+		// this ensures that the section will always be used if
+		// contains `data` substring regardless of their permissions.
+		return true;
+	}
+
+	return section->perm && !(section->perm & RZ_PERM_X);
+}
```
This will be replaced by the new function in https://github.com/rizinorg/rizin/pull/2878 PR

XVilka reviewed (approved):
Not a fan of such nestiness, but LGTM until the BB iterator is done.

wargio committed:
Removing dead code in esil.

Before this fix, rz_analysis_archinfo(RZ_ANALYSIS_ARCHINFO_DATA_ALIGN) always returned -1
because the RZ_ANALYSIS_ARCHINFO_DATA_ALIGN value was not set in the switch;
when the code was refactored, this broke esil because several checks that were always set
as true, now were actually correctly executed.
To resolve this the dead code has been removed.

wargio committed:
Fixed OOB read in 6502 analysis plugin.

wargio commented:
Before this fix, `rz_analysis_archinfo(RZ_ANALYSIS_ARCHINFO_DATA_ALIGN)` always returned `-1`, because the `RZ_ANALYSIS_ARCHINFO_DATA_ALIGN` value was not set in the switch; when the code was refactored, this broke `esil` because several checks that were always true, but now they are actually correctly executed.
To resolve this the dead code has been removed.

XVilka commented on librz/analysis/analysis.c:
```diff
@@ -446,18 +446,24 @@ RZ_API void rz_analysis_purge(RzAnalysis *analysis) {
 
-RZ_API int rz_analysis_archinfo(RzAnalysis *analysis, int query) {
-	rz_return_val_if_fail(analysis, -1);
+RZ_API int rz_analysis_archinfo(RzAnalysis *analysis, RzAnalysisInfoType query) {
```
Please add Doxygen what exactly this does

XVilka commented on librz/core/canalysis.c:
```diff
@@ -5755,6 +5755,105 @@ RZ_IPI bool rz_core_analysis_var_rename(RzCore *core, const char *name, const ch
 }
 
+static bool is_in_data_map(RzCore *core, const ut64 address) {
+	if (address < 256 || address == UT64_MAX) {
```
Why? It looks somewhat arbitrary. At least make it a define or something?

ret2libc replied:
I would remove it completely, if possible. Otherwise, comment it clearly as well

wargio replied:
No, this is not useless. there are a lot of false positives when it comes to addresses lower than 256, just imagine for example:
```
@@ --1,14 +-1,14 @@
 |       ,=< 0x000113bd      je    0x1144e
-|       |   0x000113c3      mov   esi, 0x2f                            ; '/' ; int c
+|       |   0x000113c3      mov   esi, 0x2f                            ; data.0000002f ; int c
 |       |   0x000113c8      call  sym.imp.strrchr                      ; char *strrchr(const char *s, int c)
 |       |   0x000113cd      test  rax, rax
 |      ,==< 0x000113d0      je    0x11424
 |      ||   0x000113d2      lea   rdx, [rax + 1]
 |      ||   0x000113d6      mov   rcx, rdx
 |      ||   0x000113d9      sub   rcx, rbx
-|      ||   0x000113dc      cmp   rcx, 6
+|      ||   0x000113dc      cmp   rcx, 6                               ; data.00000006
 |     ,===< 0x000113e0      jle   0x11424
 |     |||   0x000113e2      lea   rsi, [rax - 6]
 |     |||   0x000113e6      mov   ecx, 7
```

XVilka replied:
Then please add a comment.

wargio replied:
i did.

XVilka commented on librz/core/canalysis.c:
```diff
@@ -5755,6 +5755,105 @@ RZ_IPI bool rz_core_analysis_var_rename(RzCore *core, const char *name, const ch
+			}
+
+			for (ut32 i = 0; i < block->size;) {
+				ut64 pc = block->addr + i;
```
I would extract the body of this function into a separate function.

XVilka commented on librz/include/rz_analysis.h:
```diff
@@ -95,10 +95,15 @@ enum {
-#define RZ_ANALYSIS_ARCHINFO_MAX_OP_SIZE 1
-#define RZ_ANALYSIS_ARCHINFO_ALIGN       2
-#define RZ_ANALYSIS_ARCHINFO_DATA_ALIGN  4
+typedef enum rz_analysis_arch_info_type_t {
```
Please add Doxygen for these

XVilka reviewed (changes requested):
Logic looks good, please add a few tests also with this option turned off.

ret2libc commented:
I will review as well.

ret2libc commented on test/db/analysis/x86_32:
```diff
@@ -42,7 +42,7 @@ aap
 EOF
 EXPECT=<<EOF
-0x00401e80 fcn.00401e80
+0x00401e80 data.00401e80
```
This seems wrong

wargio replied:
no it is not. i checked this and is correct. the additional code adds the reference, then aap adds the function which is already named (because the flag exists) thus keeping the name.

ret2libc commented on test/db/formats/elf/elf-riscv64:
```diff
@@ -422,7 +422,7 @@ laddr    0x00000000
 machine  RISC V
 maxopsz  6
-minopsz  2
+minopsz  4
```
Is this alright? Just asking, I know nothing about RISC V arch.

wargio replied:
yes it is, it was wrong before.

ret2libc commented on librz/core/canalysis.c:
```diff
@@ -5755,6 +5755,105 @@ RZ_IPI bool rz_core_analysis_var_rename(RzCore *core, const char *name, const ch
+			if (!bytes) {
+				RZ_LOG_ERROR("Failed allocate basic block bytes buffer\n");
+				goto end;
+			} else if (0 > rz_io_nread_at(core->io, block->addr, bytes, block->size)) {
```
```suggestion
			}
			if (0 > rz_io_nread_at(core->io, block->addr, bytes, block->size)) {
```

ret2libc commented:
> Before this fix, `rz_analysis_archinfo(RZ_ANALYSIS_ARCHINFO_DATA_ALIGN)` always returned `-1`, because the `RZ_ANALYSIS_ARCHINFO_DATA_ALIGN` value was not set in the switch; when the code was refactored, this broke `esil` because several checks that were always true, but now they are actually correctly executed. To resolve this the dead code has been removed.

Are you sure the code was dead? I see 
```c
static int archinfo(RzAnalysis *analysis, int q) {
	if (q == RZ_ANALYSIS_ARCHINFO_DATA_ALIGN) {
		return 4;
	}
```
in analysis_arm_cs.c

wargio commented:
> > Before this fix, `rz_analysis_archinfo(RZ_ANALYSIS_ARCHINFO_DATA_ALIGN)` always returned `-1`, because the `RZ_ANALYSIS_ARCHINFO_DATA_ALIGN` value was not set in the switch; when the code was refactored, this broke `esil` because several checks that were always true, but now they are actually correctly executed. To resolve this the dead code has been removed.
> 
> Are you sure the code was dead? I see
> 
> ```c
> static int archinfo(RzAnalysis *analysis, int q) {
> 	if (q == RZ_ANALYSIS_ARCHINFO_DATA_ALIGN) {
> 		return 4;
> 	}
> ```
> 
> in analysis_arm_cs.c

yes it is dead code because of this:
```c

RZ_API int rz_analysis_archinfo(RzAnalysis *analysis, int query) {
	rz_return_val_if_fail(analysis, -1);
	switch (query) {
	case RZ_ANALYSIS_ARCHINFO_MIN_OP_SIZE:
	case RZ_ANALYSIS_ARCHINFO_MAX_OP_SIZE:
	case RZ_ANALYSIS_ARCHINFO_ALIGN:
		if (analysis->cur && analysis->cur->archinfo) {
			return analysis->cur->archinfo(analysis, query);
		}
		break;
	}
	return -1;
}
```

ret2libc commented on librz/analysis/analysis.c:
```diff
@@ -446,18 +446,24 @@ RZ_API void rz_analysis_purge(RzAnalysis *analysis) {
+		// Always consume at least 1 byte
+		return value > 0 ? value : 1;
+	case RZ_ANALYSIS_ARCHINFO_CAN_USE_POINTERS:
+		// When negative we assume the architecture does use them.
```
I don't understand this... For example, amd29k returns true for `RZ_ANALYSIS_ARCHINFO_CAN_USE_POINTERS`, which is positive, why do you say negative is true?

wargio replied:
you can assume that  when a developer does not handle `RZ_ANALYSIS_ARCHINFO_CAN_USE_POINTERS` the return value can be `-1`, but we always want to return a boolean, thus i prefer to handle that case even tho -1 is "true"

XVilka reviewed (changes requested):
LGTM, apart from two missing points:
- Comment for that function
- Tests with this feature disabled

wargio commented:
> LGTM, apart from two missing points:
> 
>     * Comment for that function
> 
>     * Tests with this feature disabled

I do not know yet how i should write the test for disabling the option, but what i can tell you is there is the comment.

wargio commented on librz/core/canalysis.c:
```diff
@@ -5755,6 +5755,121 @@ RZ_IPI bool rz_core_analysis_var_rename(RzCore *core, const char *name, const ch
 
+static bool is_in_data_map(RzCore *core, const ut64 address) {
+	if (address < 256 || address == UT64_MAX) {
+		// prevents a lot of false positives when it comes
+		// to invalid addresses, especially in x86 arch with
+		// mov and jmp instructions.
+		return false;
+	}
```
@XVilka

XVilka commented:
@wargio looks like it causes two fuzzing tests timeout:
```

[XX] TIMEOUT /Users/runner/work/rizin/rizin/test/bins/fuzzed <fuzz> /Users/runner/work/rizin/rizin/test/bins/fuzzed/a5ab54823e83703854378cb5bc1f587a
RZ_NOPLUGINS=1 /Users/runner/bin/rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc aaa /Users/runner/work/rizin/rizin/test/bins/fuzzed/a5ab54823e83703854378cb5bc1f587a
-- stdout

-- stderr
Copyright: SEGA MEGASIS    (C)2015 James L.
DomesticName: Liquid Space Dodger V3                          
OverseasName: Liquid Space Dodger V3                          
ProductCode: GM 01234567-89
Checksum: 0xadde
Peripherials: J               
SramCode:             
ModemCode:             
CountryCode: 
[ ] Analyze all flags starting with sym. and entry0 (aa)
[
[x] Analyze all flags starting with sym. and entry0 (aa)

[ ] Analyze function calls
[
[x] Analyze function calls
[ ] Analyze len bytes of instructions for references
[
[x] Analyze len bytes of instructions for references
[ ] Check for classes
[
[x] Check for classes

[x] Finding xrefs in noncode section with analysis.in=io.maps

[x] Analyze value pointers (aav)

[x] Value from 0x00000200 to 0x0001d4f0 (aav)

[x] 0x00000200-0x0001d4f0 in 0x200-0x1d4f0 (aav)
[ ] Emulate functions to find computed references
[
[x] Emulate functions to find computed references
[ ] Analyze local variables and arguments
[
[x] Analyze local variables and arguments
[ ] Applying signatures from sigdb
[
[x] Applied 0 FLIRT signatures via sigdb
[ ] Propagate noreturn information
[
[x] Propagate noreturn information
[ ] Resolve pointers to data sections
[
-- exit status: -1


[XX] TIMEOUT /Users/runner/work/rizin/rizin/test/bins/fuzzed <fuzz> /Users/runner/work/rizin/rizin/test/bins/fuzzed/9255cead7e2cb22611359fec200090da
RZ_NOPLUGINS=1 /Users/runner/bin/rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc aaa /Users/runner/work/rizin/rizin/test/bins/fuzzed/9255cead7e2cb22611359fec200090da
-- stdout

-- stderr
Copyright: SEGA MEGASIS    (C)2015 James L.
DomesticName: Liquid Space Dodger V3                          
OverseasName: Liquid Space Dodger V3                          
ProductCode: GM 01234567-89
Checksum: 0xadde
Peripherials: J               
SramCode:             
ModemCode:             
CountryCode: 
[ ] Analyze all flags starting with sym. and entry0 (aa)
[
[x] Analyze all flags starting with sym. and entry0 (aa)

[ ] Analyze function calls
[
[x] Analyze function calls
[ ] Analyze len bytes of instructions for references
[
[x] Analyze len bytes of instructions for references
[ ] Check for classes
[
[x] Check for classes

[x] Finding xrefs in noncode section with analysis.in=io.maps

[x] Analyze value pointers (aav)

[x] Value from 0x00000200 to 0x0001d503 (aav)

[x] 0x00000200-0x0001d503 in 0x200-0x1d503 (aav)
[ ] Emulate functions to find computed references
[
[x] Emulate functions to find computed references
[ ] Analyze local variables and arguments
[
[x] Analyze local variables and arguments
[ ] Applying signatures from sigdb
[
[x] Applied 0 FLIRT signatures via sigdb
[ ] Propagate noreturn information
[
[x] Propagate noreturn information
[ ] Resolve pointers to data sections
[
-- exit status: -1
[**]    /Users/runner/work/rizin/rizin/test/bins/fuzzed      544 OK         0 BR        2 XX        0 FX
```
